File: solutions/039.combination-sum/combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
const combinationSum = (candidates, target) => {
  let res = []

  const bfs = (temp, left, pos) => {
    if (left < 0) return

    if (left === 0) {
      res.push(temp)
      return
    }

    // can use same position again but not the position before
    for (let i = pos; i < candidates.length; i++) {
      bfs([...temp, candidates[i]], left - candidates[i], i)
    }
  }

  bfs([], target, 0)

  return res
};

File: solutions/039.combination-sum/combination-sum.py
class Solution:
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        ret = []
        candidates.sort()

        def traverse(pos, tmp, left):
            nonlocal candidates, ret

            if left < 0:
                return

            if left == 0:
                ret.append(tmp)

            for i in range(len(candidates)):
                if i < pos:
                    continue
                traverse(i, tmp + [candidates[i]], left - candidates[i])

        traverse(0, [], target)

        return ret
